cinnabar/femap.py

```diff
@@ -0,0 +1,177 @@
+    ...                                         uncertainty=0.11 * kJpm)
+    >>> # Incrementally created FEMap
+    >>> fe = FEMap()
+    >>> fe.add_measurement(experimental_result1)
+    >>> fe.add_measurement(experimental_result2)
+    >>> fe.add_measurement(calculated_result)
+    """
+    # Measurements are split between computational and experimental graphs
+    # each relative Measurement is an edge between two labels (ligands)
+    # absolute Measurements are an edge between 'NULL' and the label
+    # all edges are directed, all edges can be multiply defined
+    computational_graph: nx.MultiDiGraph
+    experimental_graph: nx.MultiDiGraph
+
+    def __init__(self):
+        self.computational_graph = nx.MultiDiGraph()
+        self.experimental_graph = nx.MultiDiGraph()
+
+    @classmethod
+    def from_csv(cls, filename):
+        """Construct from legacy csv format"""
+        data = read_csv(filename)
+
+        # unpack data dictionary
+        fe = cls()
+        for r in data['Calculated']:
+            fe.add_measurement(r)
+        for r in data['Experimental'].values():
+            fe.add_measurement(r)
+
+        return fe
+
+    def add_measurement(self, measurement: Union[RelativeMeasurement, AbsoluteMeasurement]):
+        """Add new observation to FEMap, modifies the FEMap in-place
+
+        Any other attributes on the measurement are used as annotations
+
+        Raises
+        ------
+        ValueError : if bad type given
+        """
+        if isinstance(measurement, AbsoluteMeasurement):
+            # coerce to relative to simplify logic
+            meas_ = RelativeMeasurement(labelA='NULL',
+                                        labelB=measurement.label,
+                                        DDG=measurement.DG,
+                                        uncertainty=measurement.uncertainty,
+                                        computational=measurement.computational)
+        elif isinstance(measurement, RelativeMeasurement):
+            meas_ = measurement
+        else:
+            raise ValueError()
+
+        # slurp out tasty data, anything but labels
+        d = dict(meas_)
+        d.pop('labelA', None)
+        d.pop('labelB', None)
+        d.pop('label', None)
+
+        if meas_.computational:
+            self.computational_graph.add_edge(meas_.labelA, meas_.labelB, **d)
+        else:
+            self.experimental_graph.add_edge(meas_.labelA, meas_.labelB, **d)
+
+    @property
+    def n_measurements(self) -> int:
+        """Total number of both experimental and computational measurements"""
+        return len(self.experimental_graph.edges) + len(self.computational_graph.edges)
+
+    @property
+    def n_ligands(self) -> int:
+        """Total number of unique ligands"""
+        # must ignore NULL sentinel node
+        exptl = self.experimental_graph.nodes - {'NULL'}
+        compt = self.computational_graph.nodes
```

Comment:
what if you have non overlapping comp & experimental nodes?

Reply:
i.e. do we have checks in place somewhere to avoid that edge case?

Reply:
see above, I don't know how `n_nodes` is defined for non-overlapping cases.